File: server/src/game/ai/modules/hand_thinning.rs
use crate::game::{ai::types::{Candidate, ModuleContext}, types::Card};

/// Prefer playing from the shape we hold the most of.
/// Shedding the dominant shape fastest reduces the chance of getting stuck.
pub fn hand_thinning(candidate: &Candidate, ctx: &ModuleContext<'_>) -> f64 {
    let Candidate::PlaySuit { shape, .. } = candidate else { return 0.0 };
    let Some(seat) = ctx.state.seats.get(ctx.seat_index) else { return 0.0 };

    seat.hand
        .iter()
        .filter(|c| matches!(c, Card::Suit { shape: s, .. } if s == shape))
        .count() as f64
}

File: app/src/lib/game/computer/modules/hand-thinning.ts
import { isSuitCard } from '../../guards.js';
import type { Candidate, ModuleContext, ScoringModule } from '../types.js';

// Prefer playing from the shape we hold the most cards of.
// Shedding the dominant shape fastest reduces the chance of getting stuck.
export const handThinning: ScoringModule = (candidate: Candidate, ctx: ModuleContext): number => {
	if (candidate.kind !== 'play-suit') return 0;

	const player = ctx.state.players[ctx.playerIndex];
	if (player === undefined) return 0;

	const shapeCount = player.hand.filter(
		(c) => isSuitCard(c) && c.shape === candidate.card.shape
	).length;

	return shapeCount;
};
